src/app/dao/tests/JoinerDAO.spec.ts

```diff
@@ -1,17 +1,483 @@
+import { PartDAO } from '../PartDAO';
+import { PartMocks } from 'src/app/domain/PartMocks.spec';
+import { JoinerMocks } from 'src/app/domain/JoinerMocks.spec';
+import { MinimalUser } from 'src/app/domain/MinimalUser';
+import { FirstPlayer, Joiner, PartStatus, PartType } from 'src/app/domain/Joiner';
+import { MGPOptional } from 'src/app/utils/MGPOptional';
+import { createUnverifiedUser, createUser, reconnectUser, signOut } from 'src/app/services/tests/ConnectedUserService.spec';
+import { UserMocks } from 'src/app/domain/UserMocks.spec';
 
 describe('JoinerDAO', () => {
 
-    let dao: JoinerDAO;
+    let partDAO: PartDAO;
+    let joinerDAO: JoinerDAO;
+
+    const CREATOR_EMAIL: string = UserMocks.CREATOR_AUTH_USER.email.get();
+    const CREATOR_NAME: string = UserMocks.CREATOR_AUTH_USER.username.get();
+
+    const CANDIDATE_EMAIL: string = UserMocks.CANDIDATE_AUTH_USER.email.get();
+    const CANDIDATE_NAME: string = UserMocks.CANDIDATE_AUTH_USER.username.get();
+
+    const MALICIOUS_EMAIL: string = 'm@licio.us';
+    const MALICIOUS_NAME: string = 'malicious';
 
     beforeEach(async() => {
         await setupEmulators();
-        dao = TestBed.inject(JoinerDAO);
+        joinerDAO = TestBed.inject(JoinerDAO);
+        partDAO = TestBed.inject(PartDAO);
     });
     it('should be created', () => {
-        expect(dao).toBeTruthy();
+        expect(joinerDAO).toBeTruthy();
+    });
+
+    describe('for non-verified users', () => {
+        it('should forbid to create a joiner', async() => {
+            // Given an existing part.
+            // (Note that the non-verified user in practice can never have created the corresponding part,
+            // but this is an extra check just in case)
+            await createUser(CREATOR_EMAIL, CREATOR_NAME);
+            const partId: string = await partDAO.create(PartMocks.INITIAL);
+            await signOut();
+
+            // and a non-verified user
+            const nonVerifiedUser: MinimalUser = await createUnverifiedUser(MALICIOUS_EMAIL, MALICIOUS_NAME);
+
+            // When the user tries to create the joiner
+            const result: Promise<void> = joinerDAO.set(partId, { ...JoinerMocks.INITIAL, nonVerifiedUser });
+
+            // Then it should fail
+            await expectFirebasePermissionDenied(result);
+        });
+        it('should forbid to add themselves to the candidates', async() => {
+            // Given a part, and a non-verified user
+            const creator: MinimalUser = await createUser(CREATOR_EMAIL, CREATOR_NAME);
+            const partId: string = await partDAO.create(PartMocks.INITIAL);
+            const joiner: Joiner = { ...JoinerMocks.INITIAL, creator };
+            await joinerDAO.set(partId, joiner);
+            await signOut();
+
+            const nonVerifiedUser: MinimalUser = await createUnverifiedUser(MALICIOUS_EMAIL, MALICIOUS_NAME);
+            // When adding the non-verified user to the candidates
```

Comment:
since this "when" take into account WHO is logged, perhaps mentionning "When *the user add himself* to the candidates" ?